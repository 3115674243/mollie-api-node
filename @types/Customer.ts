declare namespace Mollie {
  interface Customer {
    name?: string;
    email?: string;
    locale: Locale;
    metadata?: any;

    // Access token parameters
    testmode?: boolean;
  }
}
