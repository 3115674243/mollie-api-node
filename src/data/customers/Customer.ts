import { ApiMode, Links, Url, Locale, PaymentMethod } from '../global';
import Model from '../Model';
import commonHelpers from '../commonHelpers';
import Seal from '../../types/Seal';

/**
 * Customer Response object.
 *
 * @see https://docs.mollie.com/reference/v2/customers-api/get-customer
 */
export interface CustomerData extends Model<'customer'> {
  mode: ApiMode;
  name: string;
  email: string;
  locale: Locale;
  recentlyUsedMethods: PaymentMethod[];
  metadata: Record<string, string>;
  createdAt: string;
  _links: CustomerLinks;
}

type Customer = Seal<CustomerData, typeof commonHelpers>;

export default Customer;

/**
 * Customer _links object
 *
 * @param mandates - Mandates list
 * @param subscriptions - Subscriptions list
 * @param payments - Payments list
 */
export interface CustomerLinks extends Links {
  mandates: Url;
  subscriptions: Url;
  payments: Url;
}

export function injectPrototypes(input: CustomerData): Customer {
  return Object.assign(Object.create(commonHelpers), input);
}
