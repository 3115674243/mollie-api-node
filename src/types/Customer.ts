declare namespace Mollie {
  interface ICustomer {
    resource: string;
    id: string;
    mode: ApiMode;
    name: string;
    email: string;
    locale: Locale;
    recentlyUsedMethods: Array<Method>;
    metadata: any;
    createdAt: string;
    _links: ILinks;

    // Access token parameters
    testmode?: boolean;
  }

  namespace Customer {
    namespace Params {
      interface ICreate {
        /** @param name - The full name of the customer.  */
        name?: string;
        /** @param email - The email address of the customer. */
        email?: string;
        /** @param locale - Allows you to preset the language
         *                  to be used in the hosted payment
         *                  pages shown to the consumer.
         *                  When this parameter is not provided,
         *                  the browser language will be used
         *                  instead in the payment flow
         *                  (which is usually more accurate).
         */
        locale?: Locale;
        /** @param metadata - Provide any data you like,
         *                    and we will save the data alongside
         *                    the customer.
         *                    Whenever you fetch the customer with
         *                    our API, weâ€™ll also include the metadata.
         *                    You can use up to 1kB of JSON.
         */
        metadata?: any;

        // Access token parameters
        /** @param testmode - Set this to true to create a test mode customer. */
        testmode?: boolean;
      }

      interface IGet {
        // Access token parameters
        testmode?: boolean;
      }

      interface IUpdate {
        name?: string;
        email?: string;
        locale?: Locale;
        metadata?: any;

        // Access token parameters
        testmode?: boolean;
      }

      interface IDelete {
        // Access token parameters
        testmode?: boolean;
      }

      interface IList {
        from?: string;
        limit?: number;
      }
    }

    namespace Callback {
      type Create = (error: any, customer: ICustomer) => void;
      type Get = (error: any, customer: ICustomer) => void;
      type Update = (error: any, customer: ICustomer) => void;
      type Delete = (error: any, success: boolean) => void;
      type List = (error: any, customers: Mollie.List<ICustomer>) => void;
    }
  }

  namespace CustomerPayments {
    namespace Params {
      interface ICreate {
        customerId: string;

        name?: string;
        email?: string;
        locale: Locale;
        metadata?: any;

        // Access token parameters
        testmode?: boolean;
      }

      interface IList {
        customerId: string;

        from?: string;
        limit?: number;

        // Access token parameters
        testmode?: boolean;
      }
    }

    namespace Callback {
      type Create = (error: any, customer: ICustomer) => void;
      type List = (error: any, customers: Mollie.List<ICustomer>) => void;
    }
  }
}
