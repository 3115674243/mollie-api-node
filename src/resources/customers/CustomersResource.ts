import { CreateParameters, DeleteParameters, GetParameters, ListParameters, UpdateParameters } from './parameters';
import ApiError from '../../errors/ApiError';
import Callback from '../../types/Callback';
import Customer, { CustomerData, injectPrototypes } from '../../data/customers/Customer';
import List from '../../data/list/List';
import Resource from '../Resource';
import checkId from '../../plumbing/checkId';
import renege from '../../plumbing/renege';

/**
 * The `Customers` resource
 */
export default class CustomersResource extends Resource<CustomerData, Customer> {
  protected getResourceUrl(): string {
    return 'customers';
  }

  protected injectPrototypes = injectPrototypes;

  /**
   * List Customers.
   *
   * @since 2.0.0
   *
   * @see https://docs.mollie.com/reference/v2/customers-api/list-customers
   *
   * @public ✓ This method is part of the public API
   *
   * @alias list
   */
  public all: CustomersResource['list'] = this.list;
  /**
   * List Customers.
   *
   * @since 3.0.0
   *
   * @see https://docs.mollie.com/reference/v2/customers-api/list-customers
   *
   * @public ✓ This method is part of the public API
   *
   * @alias list
   */
  public page: CustomersResource['list'] = this.list;
  /**
   * Delete a Customer.
   *
   * @since 2.0.0
   *
   * @see https://docs.mollie.com/reference/v2/customers-api/delete-customer
   *
   * @public ✓ This method is part of the public API
   *
   * @alias delete
   */
  public cancel: CustomersResource['delete'] = this.delete;

  /**
   * Creates a simple minimal representation of a customer in the Mollie API
   * to use for the {@link https://www.mollie.com/en/checkout Mollie Checkout}
   * and Recurring features.
   * These customers will appear in your
   * {@link https://www.mollie.com/dashboard/ Mollie Dashboard}
   * where you can manage their details,
   * and also see their payments and subscriptions.
   *
   * @param params - Create customer parameters
   * @param cb - (DEPRECATED SINCE 3.0.0) Callback function, can be used instead of the returned `Promise` object
   *
   * @returns The newly created customer object
   *
   * @since 2.0.0
   *
   * @see https://docs.mollie.com/reference/v2/customers-api/create-customer
   *
   * @public ✓ This method is part of the public API
   */
  public create(parameters: CreateParameters): Promise<Customer>;
  public create(parameters: CreateParameters, callback: Callback<Customer>): void;
  public create(parameters: CreateParameters) {
    if (renege(this, this.create, ...arguments)) return;
    return this.network.post(this.getResourceUrl(), parameters);
  }

  /**
   * Retrieve a single customer by its ID
   *
   * @param id - Customer ID
   * @param params - Retrieve customer parameters
   *                 (DEPRECATED SINCE 3.0.0) Can also be a callback function
   * @param cb - (DEPRECATED SINCE 3.0.0) Callback function, can be used instead of the returned `Promise` object
   *
   * @returns Customer object
   *
   * @since 2.0.0
   *
   * @see https://docs.mollie.com/reference/v2/customers-api/get-customer
   *
   * @public ✓ This method is part of the public API
   */
  public get(id: string, parameters?: GetParameters): Promise<Customer>;
  public get(id: string, parameters: GetParameters, callback: Callback<Customer>): void;
  public get(id: string, parameters?: GetParameters) {
    if (renege(this, this.get, ...arguments)) return;
    if (!checkId(id, 'customer')) {
      throw new ApiError('The customer id is invalid');
    }
    return this.network.get(`${this.getResourceUrl()}/${id}`, parameters);
  }

  /**
   * List customers
   *
   * @param params - List customer parameters
   *                 (DEPRECATED SINCE 3.0.0) Can also be a callback function
   * @param cb - (DEPRECATED SINCE 3.0.0) Callback function, can be used instead of the returned `Promise` object
   *
   * @returns
   *
   * @since 3.0.0
   *
   * @see https://docs.mollie.com/reference/v2/customers-api/list-customers
   *
   * @public ✓ This method is part of the public API
   */
  public list(parameters?: ListParameters): Promise<List<Customer>>;
  public list(parameters: ListParameters, callback: Callback<List<Customer>>): void;
  public list(parameters: ListParameters = {}) {
    if (renege(this, this.list, ...arguments)) return;
    return this.network.list(this.getResourceUrl(), 'customers', parameters).then(result => this.injectPaginationHelpers(result, this.list, parameters));
  }

  /**
   * Update a customer
   *
   * @param id - Customer ID
   * @param params - Update customer parameters
   *                 (DEPRECATED SINCE 3.0.0) Can also be a callback function
   * @param cb - (DEPRECATED SINCE 3.0.0) Callback function, can be used instead of the returned `Promise` object
   *
   * @returns The updated Customer object
   *
   * @since 2.0.0
   *
   * @see https://docs.mollie.com/reference/v2/customers-api/update-customer
   *
   * @public ✓ This method is part of the public API
   */
  public update(id: string, parameters: UpdateParameters): Promise<Customer>;
  public update(id: string, parameters: UpdateParameters, callback: Callback<Customer>): void;
  public update(id: string, parameters: UpdateParameters) {
    if (renege(this, this.update, ...arguments)) return;
    if (!checkId(id, 'customer')) {
      throw new ApiError('The customer id is invalid');
    }
    return this.network.patch(`${this.getResourceUrl()}/${id}`, parameters);
  }

  /**
   * Delete a customer
   *
   * @param id - Customer ID
   * @param params - Delete customer parameters
   *                 (DEPRECATED SINCE 3.0.0) Can also be a callback function
   * @param cb - (DEPRECATED SINCE 3.0.0) Callback function, can be used instead of the returned `Promise` object
   *
   * @returns Success status
   *
   * @since 2.0.0
   *
   * @see https://docs.mollie.com/reference/v2/customers-api/delete-customer
   *
   * @public ✓ This method is part of the public API
   */
  public delete(id: string, parameters?: DeleteParameters): Promise<true>;
  public delete(id: string, parameters: DeleteParameters, callback: Callback<true>): void;
  public delete(id: string, parameters?: DeleteParameters) {
    if (renege(this, this.delete, ...arguments)) return;
    if (!checkId(id, 'customer')) {
      throw new ApiError('The customer id is invalid');
    }
    return this.network.delete<true>(`${this.getResourceUrl()}/${id}`);
  }
}
