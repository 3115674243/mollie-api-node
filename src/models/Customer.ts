import Model from '../model';
import { ICustomer } from '../types/customer';

/**
 * The `Customer` model
 *
 * {@link ICustomer}
 */
export default class Customer extends Model implements ICustomer {
  public static resourcePrefix = 'cst_';

  public resource = 'customer';
  public id = null;
  public mode = null;
  public name = null;
  public email = null;
  public locale = null;
  public metadata = null;
  public recentlyUsedMethods = null;
  public createdAt = null;
  public _links = {
    self: null,
    documentation: null,
    mandates: null,
    subscriptions: null,
    payments: null,
  };

  // Access token parameters
  public testmode?: boolean;

  /**
   * Customer constructor
   *
   * @public âœ“ This method is part of the public API
   */
  public constructor(props?: Partial<ICustomer>) {
    super();

    Object.assign(this, props);
  }
}
