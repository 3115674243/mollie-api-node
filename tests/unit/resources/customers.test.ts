import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import Customers from '../../../src/resources/customers';
import Customer from '../../../src/models/Customer';

import response from '../__stubs__/customers.json';
import CustomersResource from '../../../src/resources/customers';

const mock = new MockAdapter(axios);

const props = {
  id: 'cst_kEn1PlbGa',
  name: 'Customer A',
  email: 'customer@example.org',
};

describe('customers', () => {
  let customers: CustomersResource;
  beforeEach(() => {
    customers = new Customers(axios.create());
  });

  it('should have a resource name and model', () => {
    const customer = new Customers(null);
    expect(customer.resource).toBe('customers');
    expect(customer.model).toBe(Customer);
  });

  describe('.create()', () => {
    mock.onPost('/customers').reply(200, response._embedded.customers[0]);

    it('should return a customer instance', done => {
      customers.create(props)
        .then(result => {
          expect(result).toBeInstanceOf(Customer);
          expect(result.name).toBe(props.name);
          expect(result.email).toBe(props.email);
          done();
        })
        .catch(error => {
          expect(error).toBeUndefined();
          done();
        });
    });

    it('should work with a callback', done => {
      customers.create(props, (err, result) => {
        expect(err).toBeNull();
        expect(result).toBeInstanceOf(Customer);
        expect(result.name).toBe(props.name);
        expect(result.email).toBe(props.email);
        done();
      });
    });
  });

  describe('.get()', () => {
    const error = { error: { message: 'The customer id is invalid' } };

    mock.onGet(`/customers/${props.id}`).reply(200, response._embedded.customers[0]);
    mock.onGet('/customers/foo').reply(500, error);

    it('should return a customer instance', done => {
      customers.get(props.id)
        .then(result => {
          expect(result).toBeInstanceOf(Customer);
          expect(result.id).toBe(props.id);
          done();
        })
        .catch(error => {
          expect(error).toBeUndefined();
          done();
        });
    });

    it('should work with a callback', done => {
      customers.get(props.id, (err, customer) => {
        expect(err).toBeNull();
        expect(customer).toBeInstanceOf(Customer);
        expect(customer.id).toBe(props.id);
        done();
      });
    });

    it('should return an error for non-existing IDs', () => {
      customers
        .get('foo')
        .then(() => {
          throw new Error('Should reject');
        })
        .catch(err => {
          expect(err).toEqual(error);
        });
    });

    it('should return an error with a callback for non-existing IDs', done => {
      customers.get('foo', (err, result) => {
        expect(err).toEqual(error);
        expect(result).toBeUndefined();
        done();
      });
    });
  });

  describe('.update()', () => {
    const error = { error: { message: 'The customer id is invalid' } };

    mock.onPost(`/customers/${props.id}`).reply(200, response._embedded.customers[0]);
    mock.onPost('/customers/foo').reply(500, error);

    it('should return a customer instance', done => {
      customers.update(props.id, props)
        .then(result => {
          expect(result).toBeInstanceOf(Customer);
          expect(result).toMatchSnapshot();
          done();
        })
        .catch(err => {
          expect(err).toBeUndefined();
          done();
        });
    });

    it('should work with a callback', done => {
      customers.update(props.id, props, (err, result) => {
        expect(err).toBeNull();
        expect(result).toBeInstanceOf(Customer);
        expect(result).toMatchSnapshot();
        done();
      });
    });

    it('should return an error for non-existing IDs', () => {
      // @ts-ignore -- this is on purpose
      customers
        .update('foo')
        .then(() => {
          throw new Error('Should reject');
        })
        .catch(err => {
          expect(err).toEqual(error);
        });
    });

    it('should return an error with a callback for non-existing IDs', done => {
      // @ts-ignore -- this is on purpose
      customers.update('foo', (err, result) => {
        expect(err).toEqual(error);
        expect(result).toBeUndefined();
        done();
      });
    });
  });

  describe('.list()', () => {
    mock.onGet('/customers').reply(200, response);

    it('should return a list of all customers', () => {
      customers.list().then(result => {
        expect(result).toBeInstanceOf(Array);
        expect(result).toHaveProperty('links');
        expect(result).toMatchSnapshot();
      });
    });

    it('should work with a callback', done => {
      customers.list((err, result) => {
        expect(err).toBeNull();
        expect(result).toBeInstanceOf(Array);
        expect(result).toHaveProperty('links');
        expect(result).toMatchSnapshot();
        done();
      });
    });
  });
});
