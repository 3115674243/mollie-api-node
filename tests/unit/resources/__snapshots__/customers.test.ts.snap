// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customers .list() should return a list of all customers 1`] = `
Array [
  Customer {
    "_links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/reference/v2/customers-api/get-customer",
        "type": "text/html",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/customers/cst_kEn1PlbGa",
        "type": "application/hal+json",
      },
    },
    "createdAt": "2018-04-06T13:23:21.0Z",
    "data": Object {
      "_links": Object {
        "documentation": Object {
          "href": "https://docs.mollie.com/reference/v2/customers-api/get-customer",
          "type": "text/html",
        },
        "self": Object {
          "href": "https://api.mollie.com/v2/customers/cst_kEn1PlbGa",
          "type": "application/hal+json",
        },
      },
      "createdAt": "2018-04-06T13:23:21.0Z",
      "email": "customer@example.org",
      "id": "cst_kEn1PlbGa",
      "locale": "nl_NL",
      "metadata": null,
      "mode": "test",
      "name": "Customer A",
      "recentlyUsedMethods": Array [
        "creditcard",
        "ideal",
      ],
      "resource": "customer",
    },
    "email": "customer@example.org",
    "id": "cst_kEn1PlbGa",
    "locale": "nl_NL",
    "metadata": null,
    "mode": "test",
    "name": "Customer A",
    "recentlyUsedMethods": Array [
      "creditcard",
      "ideal",
    ],
    "resource": "customer",
  },
]
`;

exports[`customers .list() should work with a callback 1`] = `
Array [
  Customer {
    "_links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/reference/v2/customers-api/get-customer",
        "type": "text/html",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/customers/cst_kEn1PlbGa",
        "type": "application/hal+json",
      },
    },
    "createdAt": "2018-04-06T13:23:21.0Z",
    "data": Object {
      "_links": Object {
        "documentation": Object {
          "href": "https://docs.mollie.com/reference/v2/customers-api/get-customer",
          "type": "text/html",
        },
        "self": Object {
          "href": "https://api.mollie.com/v2/customers/cst_kEn1PlbGa",
          "type": "application/hal+json",
        },
      },
      "createdAt": "2018-04-06T13:23:21.0Z",
      "email": "customer@example.org",
      "id": "cst_kEn1PlbGa",
      "locale": "nl_NL",
      "metadata": null,
      "mode": "test",
      "name": "Customer A",
      "recentlyUsedMethods": Array [
        "creditcard",
        "ideal",
      ],
      "resource": "customer",
    },
    "email": "customer@example.org",
    "id": "cst_kEn1PlbGa",
    "locale": "nl_NL",
    "metadata": null,
    "mode": "test",
    "name": "Customer A",
    "recentlyUsedMethods": Array [
      "creditcard",
      "ideal",
    ],
    "resource": "customer",
  },
]
`;

exports[`customers .update() should return a customer instance 1`] = `
Customer {
  "_links": Object {
    "documentation": null,
    "mandates": null,
    "payments": null,
    "self": null,
    "subscriptions": null,
  },
  "config": Object {
    "data": "{\\"id\\":\\"cst_kEn1PlbGa\\",\\"name\\":\\"Customer A\\",\\"email\\":\\"customer@example.org\\"}",
    "headers": Object {
      "Accept": "application/json, text/plain, */*",
      "Content-Type": "application/json;charset=utf-8",
    },
    "maxContentLength": -1,
    "method": "post",
    "timeout": 0,
    "transformRequest": Object {
      "0": [Function],
    },
    "transformResponse": Object {
      "0": [Function],
    },
    "url": "customers/cst_kEn1PlbGa",
    "validateStatus": [Function],
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
  },
  "createdAt": null,
  "data": Object {
    "_links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/reference/v2/customers-api/get-customer",
        "type": "text/html",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/customers/cst_kEn1PlbGa",
        "type": "application/hal+json",
      },
    },
    "createdAt": "2018-04-06T13:23:21.0Z",
    "email": "customer@example.org",
    "id": "cst_kEn1PlbGa",
    "locale": "nl_NL",
    "metadata": null,
    "mode": "test",
    "name": "Customer A",
    "recentlyUsedMethods": Array [
      "creditcard",
      "ideal",
    ],
    "resource": "customer",
  },
  "email": null,
  "headers": undefined,
  "id": null,
  "locale": null,
  "metadata": null,
  "mode": null,
  "name": null,
  "recentlyUsedMethods": null,
  "resource": "customer",
  "status": 200,
}
`;

exports[`customers .update() should work with a callback 1`] = `
Customer {
  "_links": Object {
    "documentation": null,
    "mandates": null,
    "payments": null,
    "self": null,
    "subscriptions": null,
  },
  "config": Object {
    "data": "{\\"id\\":\\"cst_kEn1PlbGa\\",\\"name\\":\\"Customer A\\",\\"email\\":\\"customer@example.org\\"}",
    "headers": Object {
      "Accept": "application/json, text/plain, */*",
      "Content-Type": "application/json;charset=utf-8",
    },
    "maxContentLength": -1,
    "method": "post",
    "timeout": 0,
    "transformRequest": Object {
      "0": [Function],
    },
    "transformResponse": Object {
      "0": [Function],
    },
    "url": "customers/cst_kEn1PlbGa",
    "validateStatus": [Function],
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
  },
  "createdAt": null,
  "data": Object {
    "_links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/reference/v2/customers-api/get-customer",
        "type": "text/html",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/customers/cst_kEn1PlbGa",
        "type": "application/hal+json",
      },
    },
    "createdAt": "2018-04-06T13:23:21.0Z",
    "email": "customer@example.org",
    "id": "cst_kEn1PlbGa",
    "locale": "nl_NL",
    "metadata": null,
    "mode": "test",
    "name": "Customer A",
    "recentlyUsedMethods": Array [
      "creditcard",
      "ideal",
    ],
    "resource": "customer",
  },
  "email": null,
  "headers": undefined,
  "id": null,
  "locale": null,
  "metadata": null,
  "mode": null,
  "name": null,
  "recentlyUsedMethods": null,
  "resource": "customer",
  "status": 200,
}
`;
